storage raft retry_join does not support public_v6, incorrectly parses IP addresses as URLs

<!-- Please reserve GitHub issues for bug reports and feature requests.

For questions, the best place to get answers is on our [discussion forum](https://discuss.hashicorp.com/c/vault), as they will get more visibility from experienced users than the issue tracker.

Please note: We take Vault's security and our users' trust very seriously. If you believe you have found a security issue in Vault, please responsibly disclose by contacting us at security@hashicorp.com. Our PGP key is available at [our security page](https://www.hashicorp.com/security/).

-->

**Describe the bug**
When retry_join is set to use `addr_type=public_v6`, Vault is unable to parse the IP addresses provided by go-discover. Vault uses url.Parse on [addrs](https://github.com/hashicorp/vault/blob/v1.7.3/vault/raft.go#L958), but these are IP addresses and not URLs. 

This appears to be a mismatch between go-discover's [Addrs method ](https://pkg.go.dev/github.com/hashicorp/go-discover#Discover.Addrs)and Vault's use of the values. go-discover returns IP addresses, not URLs. Attempting to use url.Parse on these values returns interesting edge-cases particularly with IPv6. I constructed a demo of this behavior: https://play.golang.org/p/HgBDYVHdEDX


**To Reproduce**
Steps to reproduce the behavior:
1. Start vault

**Expected behavior**
Vault discovers and joins other nodes.

**Environment:**
* Vault Server Version (retrieve with `vault status`):
Vault v1.7.3 (5d517c864c8f10385bf65627891bc7ef55f5e827)
* Vault CLI Version (retrieve with `vault version`):
Vault v1.7.3 (5d517c864c8f10385bf65627891bc7ef55f5e827)
* Server Operating System/Architecture:
CentOS 7 x86_64

Vault server configuration file(s):

```json
    ...
    "storage": {
        "raft": {
            "node_id": "example",
            "path": "/opt/vault/data",
            "retry_join": [
                {
                    "auto_join": "provider=aws tag_key=vault_discovery tag_value=test addr_type=public_v6",
                    "auto_join_scheme": "https",
                    "leader_client_cert_file": "/opt/vault/tls/tls.crt",
                    "leader_client_key_file": "/opt/vault/tls/tls.key",
                    "leader_tls_servername": "vault-test.com"
                }
            ]
        }
    },
```

**Additional context**

Log excerpt
```
2021-08-13T13:22:18.964Z [INFO]  core: [DEBUG] discover: Using provider "aws"
2021-08-13T13:22:18.964Z [INFO]  core: [DEBUG] discover-aws: Using region= tag_key=vault_discovery tag_value=key addr_type=public_v6
2021-08-13T13:22:18.964Z [INFO]  core: [DEBUG] discover-aws: No static credentials
2021-08-13T13:22:18.964Z [INFO]  core: [DEBUG] discover-aws: Using environment variables, shared credentials or instance role
2021-08-13T13:22:18.964Z [INFO]  core: [INFO] discover-aws: Region not provided. Looking up region in metadata...
2021-08-13T13:22:18.966Z [INFO]  core: [INFO] discover-aws: Region is us-east-1
2021-08-13T13:22:18.966Z [INFO]  core: [DEBUG] discover-aws: Creating session...
2021-08-13T13:22:18.966Z [INFO]  core: [INFO] discover-aws: Filter instances with vault_discovery=key
2021-08-13T13:22:19.022Z [INFO]  core: [DEBUG] discover-aws: Found 5 reservations
2021-08-13T13:22:19.022Z [INFO]  core: [DEBUG] discover-aws: Reservation r-073c85432ce0bbbde has 1 instances
2021-08-13T13:22:19.022Z [INFO]  core: [DEBUG] discover-aws: Found instance i-0b10b0923b9f22ca8
2021-08-13T13:22:19.022Z [INFO]  core: [DEBUG] discover-aws: Instance i-0b10b0923b9f22ca8 has 1 network interfaces
2021-08-13T13:22:19.022Z [INFO]  core: [DEBUG] discover-aws: Checking NetworInterfaceId eni-050ada1ed0547f204 on Instance i-0b10b0923b9f22ca8
2021-08-13T13:22:19.022Z [INFO]  core: [INFO] discover-aws: Instance i-0b10b0923b9f22ca8 has IPv6 dead:beef:fee1:000d:29bb:49d6:d459:8e90 on NetworkInterfaceId eni-050ada1ed0547f204
2021-08-13T13:22:19.022Z [INFO]  core: [DEBUG] discover-aws: Reservation r-0d0226ca62b2d27ec has 1 instances
2021-08-13T13:22:19.022Z [INFO]  core: [DEBUG] discover-aws: Found instance i-02fdf7f83cd99f6f2
2021-08-13T13:22:19.022Z [INFO]  core: [DEBUG] discover-aws: Instance i-02fdf7f83cd99f6f2 has 1 network interfaces
2021-08-13T13:22:19.022Z [INFO]  core: [DEBUG] discover-aws: Checking NetworInterfaceId eni-0c31e32c5791965cb on Instance i-02fdf7f83cd99f6f2
2021-08-13T13:22:19.022Z [INFO]  core: [INFO] discover-aws: Instance i-02fdf7f83cd99f6f2 has IPv6 dead:beef:fee1:000d:655c:d793:73b7:2822 on NetworkInterfaceId eni-0c31e32c5791965cb
2021-08-13T13:22:19.022Z [INFO]  core: [DEBUG] discover-aws: Reservation r-0ce21babecae77bb5 has 1 instances
2021-08-13T13:22:19.022Z [INFO]  core: [DEBUG] discover-aws: Found instance i-04031366838966b7d
2021-08-13T13:22:19.022Z [INFO]  core: [DEBUG] discover-aws: Instance i-04031366838966b7d has 1 network interfaces
2021-08-13T13:22:19.022Z [INFO]  core: [DEBUG] discover-aws: Checking NetworInterfaceId eni-0d76d634d410cd744 on Instance i-04031366838966b7d
2021-08-13T13:22:19.022Z [INFO]  core: [INFO] discover-aws: Instance i-04031366838966b7d has IPv6 dead:beef:fee1:000e:3e7c:4e31:bf61:f442 on NetworkInterfaceId eni-0d76d634d410cd744
2021-08-13T13:22:19.022Z [INFO]  core: [DEBUG] discover-aws: Reservation r-0d9cd8d490dcbba11 has 1 instances
2021-08-13T13:22:19.022Z [INFO]  core: [DEBUG] discover-aws: Found instance i-00c61eabef684c0b6
2021-08-13T13:22:19.022Z [INFO]  core: [DEBUG] discover-aws: Instance i-00c61eabef684c0b6 has 1 network interfaces
2021-08-13T13:22:19.022Z [INFO]  core: [DEBUG] discover-aws: Checking NetworInterfaceId eni-063e9feaa9c2fe84f on Instance i-00c61eabef684c0b6
2021-08-13T13:22:19.022Z [INFO]  core: [INFO] discover-aws: Instance i-00c61eabef684c0b6 has IPv6 dead:beef:fee1:000f:859c:9ee9:c55d:1ffe on NetworkInterfaceId eni-063e9feaa9c2fe84f
2021-08-13T13:22:19.022Z [INFO]  core: [DEBUG] discover-aws: Reservation r-08b05b70d2fb1e8fc has 1 instances
2021-08-13T13:22:19.022Z [INFO]  core: [DEBUG] discover-aws: Found instance i-01e16cef7618985de
2021-08-13T13:22:19.022Z [INFO]  core: [DEBUG] discover-aws: Instance i-01e16cef7618985de has 1 network interfaces
2021-08-13T13:22:19.022Z [INFO]  core: [DEBUG] discover-aws: Checking NetworInterfaceId eni-0cd647c8d05bba03e on Instance i-01e16cef7618985de
2021-08-13T13:22:19.022Z [INFO]  core: [INFO] discover-aws: Instance i-01e16cef7618985de has IPv6 dead:beef:fee1:000f:ce2b:1703:3cac:1b94 on NetworkInterfaceId eni-0cd647c8d05bba03e
2021-08-13T13:22:19.022Z [INFO]  core: [DEBUG] discover-aws: Found ip addresses: [dead:beef:fee1:000d:29bb:49d6:d459:8e90 dead:beef:fee1:000d:655c:d793:73b7:2822 dead:beef:fee1:000e:3e7c:4e31:bf61:f442 dead:beef:fee1:000f:859c:9ee9:c55d:1ffe dead:beef:fee1:000f:ce2b:1703:3cac:1b94]
2021-08-13T13:22:19.022Z [ERROR] core: failed to parse discovered address: error="parse "dead:beef:fee1:000d:29bb:49d6:d459:8e90": first path segment in URL cannot contain colon"
2021-08-13T13:22:19.022Z [ERROR] core: failed to parse discovered address: error="parse "dead:beef:fee1:000d:655c:d793:73b7:2822": first path segment in URL cannot contain colon"
2021-08-13T13:22:19.022Z [ERROR] core: failed to parse discovered address: error="parse "dead:beef:fee1:000e:3e7c:4e31:bf61:f442": first path segment in URL cannot contain colon"
2021-08-13T13:22:19.022Z [ERROR] core: failed to parse discovered address: error="parse "dead:beef:fee1:000f:859c:9ee9:c55d:1ffe": first path segment in URL cannot contain colon"
2021-08-13T13:22:19.022Z [ERROR] core: failed to parse discovered address: error="parse "dead:beef:fee1:000f:ce2b:1703:3cac:1b94": first path segment in URL cannot contain colon"
2021-08-13T13:22:19.022Z [ERROR] core: failed to retry join raft cluster: retry=2s
```
Related issue:
- #9531

--- a/changelog/12366.txt
+++ b/changelog/12366.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+storage/raft: Support `addr_type=public_v6` in auto-join
+```

--- a/vault/raft.go
+++ b/vault/raft.go
@@ -947,11 +947,6 @@ func (c *Core) JoinRaftCluster(ctx context.Context, leaderInfos []*raft.LeaderJo
 				}
 
 			case leaderInfo.AutoJoin != "":
-				// go-discover returns IPs
-				addrs, err := disco.Addrs(leaderInfo.AutoJoin, c.logger.StandardLogger(nil))
-				if err != nil {
-					c.logger.Error("failed to parse addresses from auto-join metadata", "error", err)
-				}
 				scheme := leaderInfo.AutoJoinScheme
 				if scheme == "" {
 					// default to HTTPS when no scheme is provided
@@ -962,12 +957,17 @@ func (c *Core) JoinRaftCluster(ctx context.Context, leaderInfos []*raft.LeaderJo
 					// default to 8200 when no port is provided
 					port = 8200
 				}
-				for _, addr := range addrs {
-					if strings.Count(addr, ":") != 0 {
-						// ipv6, make it explicit
-						addr = fmt.Sprintf("[%s]", addr)
+				// Addrs returns either IPv4 or IPv6 address sans scheme or port
+				clusterIPs, err := disco.Addrs(leaderInfo.AutoJoin, c.logger.StandardLogger(nil))
+				if err != nil {
+					c.logger.Error("failed to parse addresses from auto-join metadata", "error", err)
+				}
+				for _, ip := range clusterIPs {
+					if strings.Count(ip, ":") >= 2 && !strings.HasPrefix(ip, "["){
+						// An IPv6 address in implicit form, however we need it in explicit form to use in a URL.
+						ip = fmt.Sprintf("[%s]", ip)
 					}
-					u := fmt.Sprintf("%s://%s:%d", scheme, addr, port)
+					u := fmt.Sprintf("%s://%s:%d", scheme, ip, port)
 					if err := joinLeader(leaderInfo, u); err != nil {
 						c.logger.Warn("join attempt failed", "error", err)
 					} else {

--- a/vault/raft.go
+++ b/vault/raft.go
@@ -947,39 +947,28 @@ func (c *Core) JoinRaftCluster(ctx context.Context, leaderInfos []*raft.LeaderJo
 				}
 
 			case leaderInfo.AutoJoin != "":
+				// go-discover returns IPs
 				addrs, err := disco.Addrs(leaderInfo.AutoJoin, c.logger.StandardLogger(nil))
 				if err != nil {
 					c.logger.Error("failed to parse addresses from auto-join metadata", "error", err)
 				}
-
+				scheme := leaderInfo.AutoJoinScheme
+				if scheme == "" {
+					// default to HTTPS when no scheme is provided
+					scheme = "https"
+				}
+				port := leaderInfo.AutoJoinPort
+				if port == 0 {
+					// default to 8200 when no port is provided
+					port = 8200
+				}
 				for _, addr := range addrs {
-					u, err := url.Parse(addr)
-					if err != nil {
-						c.logger.Error("failed to parse discovered address", "error", err)
-						continue
+					if strings.Count(addr, ":") != 0 {
+						// ipv6, make it explicit
+						addr = fmt.Sprintf("[%s]", addr)
 					}
-
-					if u.Scheme == "" {
-						scheme := leaderInfo.AutoJoinScheme
-						if scheme == "" {
-							// default to HTTPS when no scheme is provided
-							scheme = "https"
-						}
-
-						addr = fmt.Sprintf("%s://%s", scheme, addr)
-					}
-
-					if u.Port() == "" {
-						port := leaderInfo.AutoJoinPort
-						if port == 0 {
-							// default to 8200 when no port is provided
-							port = 8200
-						}
-
-						addr = fmt.Sprintf("%s:%d", addr, port)
-					}
-
-					if err := joinLeader(leaderInfo, addr); err != nil {
+					u := fmt.Sprintf("%s://%s:%d", scheme, addr, port)
+					if err := joinLeader(leaderInfo, u); err != nil {
 						c.logger.Warn("join attempt failed", "error", err)
 					} else {
 						// successfully joined leader

